services/galley/src/Galley/External/LegalHoldService.hs

```diff
@@ -167,3 +244,43 @@ makeLegalHoldServiceRequest tid reqBuilder = do
     mkReqBuilder token =
       reqBuilder
         . Bilge.header "Authorization" ("Bearer " <> toByteString' token)
+
+versionToInt :: LhApiVersion -> Int
+versionToInt V0 = 0
+versionToInt V1 = 1
+
+versionToBS :: LhApiVersion -> ByteString
+versionToBS = ("v" <>) . BS8.pack . show . versionToInt
+
+versionedPaths :: LhApiVersion -> [ByteString] -> Http.Request -> Http.Request
+versionedPaths V0 paths = Bilge.paths paths
+versionedPaths v paths = Bilge.paths (versionToBS v : paths)
+
+supportedClientVersions :: Set LhApiVersion
+supportedClientVersions = Set.fromList [minBound .. maxBound]
+
+-- | Find the highest common version between wire-server and the legalhold service.
+-- If the legalhold service does not support the `/api-version` endpoint, we assume it's `v0`.
+-- If there is no common version, we log a warning and use `v0`. This violates the fail-fast principle,
+-- but makes multiple API endpoints backwards compatible.
+negotiateVersion ::
+  ( Member (ErrorS 'LegalHoldServiceNotRegistered) r,
+    Member LegalHoldStore r,
+    Member P.TinyLog r
+  ) =>
+  TeamId ->
+  Set LhApiVersion ->
+  Sem r LhApiVersion
+negotiateVersion tid clientVersions = do
+  mSupportedServerVersions <- getLegalHoldApiVersion tid
```

Comment:
supported by whom? i just realize this was hard to read in some of the code above as well. maybe call it "serverVersion" and "clientVersion", where wire-server is the client?  or "lhVersion" and "wireVersion"?

pls ignore, just thinking out loud.

(ah, i just see that 3 lines further down you had the same idea :)